service/history/plugins/callbacks/statemachine.go

```diff
@@ -0,0 +1,220 @@
+func NewWorkflowClosedTrigger() *workflowpb.CallbackInfo_Trigger {
+	return &workflowpb.CallbackInfo_Trigger{
+		Variant: &workflowpb.CallbackInfo_Trigger_WorkflowClosed{},
+	}
+}
+
+// NewCallback creates a new callback in the STANDBY state from given params.
+func NewCallback(registrationTime *timestamppb.Timestamp, trigger *workflowpb.CallbackInfo_Trigger, cb *commonpb.Callback) Callback {
+	return Callback{
+		&persistencespb.CallbackInfo{
+			PublicInfo: &workflowpb.CallbackInfo{
+				Trigger:          trigger,
+				Callback:         cb,
+				State:            enumspb.CALLBACK_STATE_STANDBY,
+				RegistrationTime: registrationTime,
+			},
+		},
+	}
+}
+
+func (c Callback) GetState() enumspb.CallbackState {
+	return c.PublicInfo.State
+}
+
+func (c Callback) SetState(state enumspb.CallbackState) {
+	c.PublicInfo.State = state
+}
+
+func (c Callback) recordAttempt(ts time.Time) {
+	c.PublicInfo.Attempt++
+	c.PublicInfo.LastAttemptCompleteTime = timestamppb.New(ts)
+}
+
+func (c Callback) Tasks() ([]statemachines.Task, error) {
+	switch c.PublicInfo.State {
+	case enumspb.CALLBACK_STATE_BACKING_OFF:
+		return []statemachines.Task{BackoffTask{Deadline: c.PublicInfo.NextAttemptScheduleTime.AsTime()}}, nil
+	case enumspb.CALLBACK_STATE_SCHEDULED:
+		var destination string
+		switch v := c.PublicInfo.Callback.GetVariant().(type) {
+		case *commonpb.Callback_Nexus_:
+			u, err := url.Parse(c.PublicInfo.Callback.GetNexus().Url)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse URL: %v", &c) // nolint:goerr113
+			}
+			destination = u.Scheme + "://" + u.Host
+		default:
+			return nil, fmt.Errorf("unsupported callback variant %v", v) // nolint:goerr113
+		}
+
+		return []statemachines.Task{InvocationTask{Destination: destination}}, nil
+	}
+	return nil, nil
+}
+
+type stateMachineDefinition struct{}
+
+func (stateMachineDefinition) Type() statemachines.MachineType {
+	return StateMachineType
+}
+
+func (stateMachineDefinition) Deserialize(d []byte) (any, error) {
+	info := &persistencespb.CallbackInfo{}
+	return Callback{info}, proto.Unmarshal(d, info)
+}
+
+func (stateMachineDefinition) Serialize(state any) ([]byte, error) {
+	return proto.Marshal(state.(Callback).CallbackInfo)
+}
+
+func RegisterStateMachine(r *statemachines.Registry) error {
+	return r.RegisterMachine(stateMachineDefinition{})
+}
+
+// EventScheduled is triggered when the callback is meant to be scheduled for the first time - when its Trigger
+// condition is met.
+type EventScheduled struct{}
+
+var TransitionScheduled = statemachines.NewTransition(
+	[]enumspb.CallbackState{enumspb.CALLBACK_STATE_STANDBY},
+	enumspb.CALLBACK_STATE_SCHEDULED,
+	func(cb Callback, event EventScheduled) error {
+		return nil
+	},
+)
+
+// EventRescheduled is triggered when the callback is meant to be rescheduled after backing off from a previous attempt.
+type EventRescheduled struct{}
+
+var TransitionRescheduled = statemachines.NewTransition(
```

Comment:
Both of these suggestions somewhat SGTM.

Events here are actually inputs, so I don't mind changing the terminology.
I know that a lot of FSM libraries use the term event but I agree changing the terminology to avoid confusion with history events which may also be inputs.

I'll try the struct approach, I didn't go for that because it adds what I think is unnecessary boilerplate.
I do like that all the transitions are grouped somehow, especially if we want to generate diagrams from the definitions but that doesn't require a struct, it could be based on a convention.